### PR TITLE
Fix #4480 by tracing inputs before running function.

### LINKED ISSUE #4480: DCE trick for inplace tracing doesn't work
Exposed by #4395, hotpatched in #4471.

The DCE trick says that if I have `y = f(x)`, and `f` is internally implemented as `g`, it's OK to trace both `g` and `f`. Recall the tracing algorithm is:

1. enter f(x)
2. compute its result y (recursively calling other possibly tracing functions)
3. trace y = f(x) (use the mapping to figure out the identity of x, and writing in a mapping that y binds to the result JIT IR value of this trace)
4. return from f

So when you run the example above, you'll do this:

```
# suppose x is mapped to %1
enter f(x)
enter g(x)
result of g is y
trace y = g(x a.k.a. %1) (mapping y to %2)
return from g
result of f is y
trace y = f(x a.k.a. %1) (remapping y to %3)
return from f
```

and end up with a trace like this:

```
%2 = g(%1)
%3 = f(%1)
... only %3 is live, because %2 was killed from the mapping...
```

Subsequent DCE will eliminate the invocation of `g` and you'll only see `f` in the final trace.

However, if f and g are inplace functions, the machinery breaks:

```
# suppose x is mapped to %1
enter f(x)
enter g(x)
result of g is x
trace x = g(x a.k.a. %1) (remapping x to %2)
return from g
result of f is x
trace x = f(x a.k.a. %2) (remapping x to %3)
return from f
```

resulting in:

```
%2 = g(%1)
%3 = f(%2) # OOPS
```

Proper TLS tracing will solve this problem. But maybe a more robust stopgap than #4471 is to add a little bit of extra logic for inplace functions. Essentially, what we need to do is save the original Value for a function before we make the inner recursive call, and use that as the Value when we determine the arguments of tracing, rather than doing it after we do the inner recursive call (in which case `x` could have been overwritten). In fact, I originally implemented the tracer this way but @colesbury convinced me that it wasn't necessary (and it was true, in the code at the time, it was not necessary.)

@houseroad, when you fix this, please work this discussion into a comment somewhere in the code.

CC @apaszke

### PR DESCRIPTION
**Pull request by ezyang**
The DCE trick says that if I have y = f(x), and f is internally implemented as
g, it's OK to trace both g and f. Recall the tracing algorithm is:

    enter f(x)
    compute its result y
    trace y = f(x)
    return from f

So when you run the example above, you'll do this:

    # suppose x is mapped to %1
    enter f(x)
    enter g(x)
    result of g is y
    trace y = g(x a.k.a. %1) (mapping y to %2)
    return from g
    result of f is y
    trace y = f(x a.k.a. %1) (remapping y to %3)
    return from f

and end up with a trace like this:

    %2 = g(%1)
    %3 = f(%1)

... only %3 is live, because %2 was killed from the mapping...  Subsequent DCE
will eliminate the invocation of g and you'll only see f in the final trace.

However, if f and g are inplace functions, the machinery breaks:

    # suppose x is mapped to %1
    enter f(x)
    enter g(x)
    result of g is x
    trace x = g(x a.k.a. %1) (remapping x to %2)
    return from g
    result of f is x
    trace x = f(x a.k.a. %2) (remapping x to %3)
    return from f
    resulting in:

    %2 = g(%1)
    %3 = f(%2) # OOPS

This commit changes the strategy so we instead do this:

    enter f(x)
    trace f(x)
    compute its result y
    trace y = f(x)  (computed above)
    return from f

Now we get the correct Value before it is overwritten.
Here is what the new trace code looks like:

    jit::tracer::PreTraceInfo trace_info;
    if (jit::tracer::isTracing( self, index )) {
      trace_info = jit::tracer::preRecordTrace( "index_fill", { self, index } );
      setattr(trace_info.n, jit::Symbol("dim"), dim);
      setattr(trace_info.n, jit::Symbol("value"), value);
    }
    baseType->index_fill_(self_, dim, index_, value);
    increment_version(self);
    rebase_history(self, grad_fn);
    if (trace_info.state != nullptr) {
      jit::tracer::postRecordTrace( trace_info,  { self } );
    }

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by colesbury** [comment]
Everything looks reasonable to me, but I'm not very familiar with the JIT code

**Comment by soumith** [comment]
the builds are failing (expected test failure?)

```
19:12:23 ======================================================================
19:12:23 ERROR: test_nested_inplace (__main__.TestJit)
19:12:23 ----------------------------------------------------------------------
19:12:23 Traceback (most recent call last):
19:12:23   File "test_jit.py", line 1171, in test_nested_inplace
19:12:23     self.assertExpectedTrace(trace)
19:12:23   File "test_jit.py", line 70, in assertExpectedTrace
19:12:23     self.assertExpected(str(trace), *args, **kwargs)
19:12:23   File "/var/lib/jenkins/workspace/test/common.py", line 362, in assertExpected
19:12:23     "python {} {} --accept").format(munged_id, s, __main__.__file__, munged_id))
19:12:23 RuntimeError: I got this output for TestJit.test_nested_inplace:
19:12:23 
19:12:23 graph(%0 : Double(2, 2)) {
19:12:23   %1 : Double(2, 2) = threshold[threshold={0}, value={0}](%0)
19:12:23   return (%1);
19:12:23 }
19:12:23 
19:12:23 
19:12:23 No expect file exists; to accept the current output, run:
19:12:23 python test_jit.py TestJit.test_nested_inplace --accept
```

**Comment by ezyang** [comment]
Oops! forgot to git add

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -305,15 +312,15 @@ def emit_record_trace(env):
...
         tensor_args = [arg for arg in arguments if arg['simple_type'] in {'Tensor', 'TensorList'}]
         if len(tensor_args) == 0:
-            return ''
+            return ('', '')
         if base_name in DONT_RECORD_TRACE:
-            return ''
+            return ('', '')
 
         # Note [clang-802.0.42 tuple overload bug]
         # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (houseroad) [inline tools/autograd/gen_variable_type.py]: typo: emit_$ecord_trace ==> emit_record_trace

**Review by apaszke — APPROVED** [review]

**Review by houseroad — APPROVED** [review]
LGTM, thanks!

**Pull request merged**